lib/nokogiri/html/document.rb

```diff
@@ -174,6 +177,12 @@ def parse string_or_io, url = nil, encoding = nil, options = XML::ParseOptions::
 
           if string_or_io.respond_to?(:read)
             url ||= string_or_io.respond_to?(:path) ? string_or_io.path : nil
+
+            if string_or_io.is_a?(Pathname)
+              # resolve the Pathname to the file and open it as an IO object, see #2110
+              string_or_io = string_or_io.expand_path.open
```

Comment:
@tenderlove can I get your eyes on this? I'm pretty sure `File` objects autoclose when they are GCed (and I tested that), but figured it was worth asking explicitly.

Reply:
(My inclination was to throw `expand_path` in there because you can have legitimate stuff like `Pathname('~/foo.xml')` and the overhead is minimal.)

Reply:
oh btw putting the `url` assignment _after_ the `Pathname` test was deliberate because `Pathname` does not respond to `path` but `Pathname#open` will produce a `File` which _will_ respond to `path`.

Reply:
@doriantaylor I don't have any concerns about `#expand_path`, I just want to confirm with @tenderlove that the `File` object we're creating will get cleaned up and file descriptors will get closed during object finalization.

Your `url` comment makes sense -- what confused me was that after your initial patch set (which I've squashed), it was done one way in HTML::Document and the other way in XML::Document. I'll add some test coverage to make sure it's working the way we expect.